python/pyspark/sql/session.py

```diff
@@ -570,10 +570,20 @@ def _inferSchemaFromList(
         if not data:
             raise ValueError("can not infer schema from empty dataset")
         infer_dict_as_struct = self._jconf.inferDictAsStruct()
+        infer_array_from_first_element = self._jconf.legacyInferArrayTypeFromFirstElement()
```

Comment:
Do we really need a flag? in the cases where this matters, the schema would be wrong and produce an error; would that be desirable?

Reply:
Previously it was allowed to have mixed types in a python list, as long as the types could be cast to the type enforced by the schema inferred from the first element:
```python
>>> df = spark.createDataFrame([{"a": ["1", 2]}])
>>> df.show()
+------+
|     a|
+------+
|[1, 2]|
+------+
>>> df.schema
StructType(List(StructField(a,ArrayType(StringType,true),true)))
```
With this change, creating the DataFrame causes an error:
```python
>>> df = spark.createDataFrame([{"a": ["1", 2]}])
...
TypeError: Unable to infer the type of the field a.
```
Because of this change, I think it makes sense to have the behavior configurable.

Reply:
I can see the argument for casting to the widest applicable type, but I think that should be a separate discussion. Inferring the type of an array I think should be analogous to inferring a type over rows, which raises an error in this case:
```python
>>> spark.createDataFrame([{"a": "1"}, {"a": 2}])
...
TypeError: field a: Can not merge type <class 'pyspark.sql.types.StringType'> and <class 'pyspark.sql.types.LongType'>
```

Reply:
I see it fixes some cases, by somewhat 'accidentally' correctly inferring a widening type. But it does cause some common cases to start failing, when they 'accidentally' work now (your example in this thread). It feels like the half-measure isn't worth it, as it needs a whole new flag. Is it hard to just implement logic to find the closest type for everything? that code surely already exists in the code base

Reply:
I am not super familiar with this part of the code and I don't object to the change if someone else wants to merge; just seems like it would be relatively straightforward to 'really' fix it

Reply:
Yeah, that's right that it causes a behaviour change. However, the (previous) string type coercion behaviour in an element of an array is actually a mistake I believe. For example, such type coercion is not supported in regular type inference:

```python
>>> spark.createDataFrame([{"a": "1"}, {"a" :2}])
Traceback (most recent call last):
  ...
TypeError: field a: Can not merge type <class 'pyspark.sql.types.StringType'> and <class 'pyspark.sql.types.LongType'>
```

So, what this PR actually does is to match the behaviour with non-nested type inference. The switch was added for users dependent on the previous behaviour.

Reply:
I see, that is sounding reasonable then to me. I dislike flags, but, if a follow-up made the flag irrelevant, maybe that's OK.

Reply:
By the way, I think `StringType` is the only type that supports coercion: https://github.com/apache/spark/blob/master/python/pyspark/sql/types.py#L1596-L1599